### DISCUSSION TITLE
How to configurate SSL when Start with Local Source Code

### DISCUSSION THREAD

**Discussion by zyh3826** [langgenius/dify#7541] [General]
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] [FOR CHINESE USERS] 请务必使用英文提交 Issue，否则会被关闭。谢谢！:）
- [X] Please do not modify this template :) and fill in all the required fields.

### Content

I started Dify using `Start with Local Source Code`, but can't find how to configure SSL to support HTTPS, I found that using `Docker Compose` to start, there are some environments about `Nginx` that configure SSL. 
How to do this when using `Start with Local Source Code`.

**Comment by zyh3826** [comment]
I install the SSL Certification using `Tomcat`, the simple code below is work:
```python
# coding: utf-8
import ssl

import fastapi
from fastapi.responses import FileResponse
import uvicorn

app = fastapi.FastAPI()

@app.get("/")
async def root():
    return {"message": "Hello World"}

if __name__ == '__main__':
    ssl_keyfile = 'private.key'
    ssl_certfile = 'certificate.crt'
    uvicorn.run(app, host='0.0.0.0', port=8001, ssl_certfile=ssl_certfile, ssl_keyfile=ssl_keyfile)
```

but in `Dify`, after `npm run start`, can't access the server via `https`

**Comment by akramer-zibra** [comment]
The `https` connections in the docker compose setup is enabled via `nginx` used as reverse proxy and edge router. In this setup nginx terminates the tls encryption. The `web` and `api` services are only accessible via the nginx proxy. This is why those services usually do not need to have https support, because in this setup, they are not directly exposed to the internet etc.

If you want to run the web frontend based on NextJS in a local development setup, you may use `mkcert` , like it is described here https://www.restack.io/docs/nextjs-knowledge-nextjs-https-guide

**Reply by zyh3826** [reply]
this mentioned in [https://www.restack.io/docs/nextjs-knowledge-nextjs-https-guide](https://www.restack.io/docs/nextjs-knowledge-nextjs-https-guide)
```
After generating certificates, you can use them with Next.js by customizing the next.config.js file or using a custom server setup, such as Express.js, to serve your application over HTTPS.
```
I found that `next.config.js` in `web` folder, but I don't know how to customize it, because I am completely unfamiliar with NextJS. Could you help me?

BTW, if I start up Nginx link to Dify, will it worked?

**Reply by akramer-zibra** [reply]
Seems the next-config-js documentation does not know such option: https://nextjs.org/docs/pages/api-reference/next-config-js 🤔 I haven't tried this approach myself, yet. This is just some professional guessing, sorry 😅.

This other tutorial (https://vercel.com/guides/access-nextjs-localhost-https-certificate-self-signed) uses `next dev --experimental-https` command to run the frontend with enabled https. If you install nextjs on your system, like it is described in the tutorial, you should be able to use this command in the `web` folder.  

Btw: What is your goal? Do you want to change the dify code and want your changes run locally on your system?

If you only want to run dify locally on your system without any code changes of yours, the docker-compose setup should fit better to your needs. In the docker-compose setup nginx and the other services are already configured properly to work in combination: https://docs.dify.ai/getting-started/install-self-hosted/docker-compose

**Reply by zyh3826** [reply]
> Seems the next-config-js documentation does not know such option: https://nextjs.org/docs/pages/api-reference/next-config-js 🤔 I haven't tried this approach myself, yet. This is just some professional guessing, sorry 😅.
> 
> This other tutorial (https://vercel.com/guides/access-nextjs-localhost-https-certificate-self-signed) uses `next dev --experimental-https` command to run the frontend with enabled https. If you install nextjs on your system, like it is described in the tutorial, you should be able to use this command in the `web` folder.
> 
> Btw: What is your goal? Do you want to change the dify code and want your changes run locally on your system?
> 
> If you only want to run dify locally on your system without any code changes of yours, the docker-compose setup should fit better to your needs. In the docker-compose setup nginx and the other services are already configured properly to work in combination: https://docs.dify.ai/getting-started/install-self-hosted/docker-compose

We changed some code in dify, so docker-compose did not fit our needs

**Reply by akramer-zibra** [reply]
Ah, okay.

> BTW, if I start up Nginx link to Dify, will it worked?

Sure.

**Reply by zyh3826** [reply]
> Ah, okay.
> 
> > BTW, if I start up Nginx link to Dify, will it worked?
> 
> Sure.

thanks for your reply

**Comment by zyh3826** [comment]
I start the Nginx server, and "link to " Dify, the web can access via `https` but can't login. In browser, I find some info below:
![企业微信截图_1725613686618](https://github.com/user-attachments/assets/7c183c47-8744-40d7-8945-0cedaf4542b5)
![企业微信截图_17256137063669](https://github.com/user-attachments/assets/404390a5-9b6a-4a80-adfc-74b63e2f408c)


I think maybe this is because of the API configuration in Dify, but I can't confirm
Can anyone help me?

**Reply by akramer-zibra** [reply]
I don't know what those errors mean, but you can look into dify's nginx configuration, to see how you may configure it. You'll find  the `nginx` configuration files here in this folder: https://github.com/langgenius/dify/tree/main/docker/nginx

Those files are used to build different setups. The full composition is usually defined by environment variables (.env). You will see those in the configuration files. The logic how `nginx` is provisioned with the configuration files is defined in the [docker-entrypoint.sh ](https://github.com/langgenius/dify/blob/main/docker/nginx/docker-entrypoint.sh). There you'll see for example: https is optional, depending on the `${NGINX_HTTPS_ENABLED}` value. 

The default configuration folder of nginx is `/etc/nginx`. The following lines show, how the configuration files are copied there. Those files you may also need in your setup. 
https://github.com/langgenius/dify/blob/cdb08be9517f1079117a5f97c05b02f95f160632/docker/nginx/docker-entrypoint.sh#L33-L36

I think it is important to understand, that `nginx` is used as a `reverse proxy` here in this setup and **_NOT_** as a webserver many usually think of it. As a reverse proxy, `nginx` does accept requests (http or https) and will pass those request (routing) to the different dify services (web, api, ...).

The rules for this routing are defined here in the `default.conf`: https://github.com/langgenius/dify/blob/main/docker/nginx/conf.d/default.conf.template
